README.md

# Flutter Spinner Time Picker

A Flutter package providing custom time picker widgets for easy time selection in your Flutter applications.

[![pub](https://img.shields.io/pub/v/flutter_spinner_time_picker.svg)](https://pub.dev/packages/flutter_spinner_time_picker)
[![license: MIT](https://img.shields.io/badge/license-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

A Flutter package providing custom time picker widgets for easy time selection in your Flutter applications.

## Preview

### Dark Mode
![dark-spinner-time-picker](https://github.com/hosseinvejdani/flutter_spinner_time_picker/assets/18458798/fc08130c-1144-4228-8b78-7b1ffd3a1160)


## Light Mode
![light-spinner-time-picker](https://github.com/hosseinvejdani/flutter_spinner_time_picker/assets/18458798/fc2bf388-cf84-4a8d-a8fc-681dc669bee6)


## Installation

Add the following line to your `pubspec.yaml` file:

```yaml
dependencies:
  flutter_spinner_time_picker: [latest]
```

## Usage

Import the package:

```dart
import 'package:flutter_spinner_time_picker/flutter_spinner_time_picker.dart';
```

### How to use

Example usage:

```dart
TimeOfDay selectedTime = TimeOfDay.now();

void _showTimePicker() async {
  final pickedTime = await showSpinnerTimePicker(
    context,
    initTime: selectedTime,
  );

  if (pickedTime != null) {
    setState(() {
      selectedTime = pickedTime;
    });
  }
}
```


## Fully customizable package

```dart
showSpinnerTimePicker(
  context,
  title: 'Custom Time Picker',
  backgroundColor: Colors.grey[100],
  foregroundColor: Colors.black,
  titleStyle: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
  buttonStyle: ButtonStyle(
    backgroundColor: MaterialStateProperty.all(Colors.blue),
  ),
  buttonTextStyle: TextStyle(fontSize: 16, color: Colors.white),
  barrierDismissible: true,
  // ... other customization options
);

```


## Complete Example

Here's a simple example of how to use the custom time picker widgets:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_spinner_time_picker/flutter_spinner_time_picker.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Time Picker Example',
      darkTheme: ThemeData(
        primarySwatch: Colors.deepPurple,
        useMaterial3: true,
        brightness: Brightness.dark,
      ),
      theme: ThemeData(
        primarySwatch: Colors.deepPurple,
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TimeOfDay selectedTime = TimeOfDay.now();

  void _showTimePicker() async {
    final pickedTime = await showSpinnerTimePicker(
      context,
      initTime: selectedTime,
      is24HourFormat: false,
    );

    if (pickedTime != null) {
      setState(() {
        selectedTime = pickedTime;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Time Picker Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'Selected Time:',
              style: TextStyle(fontSize: 18),
            ),
            const SizedBox(height: 10),
            Text(
              '${selectedTime.hour}:${selectedTime.minute.toString().padLeft(2, '0')}',
              style: const TextStyle(fontSize: 30, fontWeight: FontWeight.bold),
            ),
            const SizedBox(height: 20),
            FilledButton(
              onPressed: _showTimePicker,
              child: const Text('Pick a Time'),
            ),
          ],
        ),
      ),
    );
  }
}
```

For more customization options and advanced usage, refer to the [API documentation](#).

## License

This package is released under the [MIT License](LICENSE).


CHANGELOG.md
## 0.2.1

* updated changelog.md file
  
## 0.2.0

* fixed bug with time picker

## 0.1.0

* removed `SpinnerNumericPicker` Widget and add `SpinnerItemPicker<T>` widget instead 

## 0.0.5

* remove  files 

## 0.0.4

* updated readme file

## 0.0.3

* updated readme file

## 0.0.2

* added screenshots

## 0.0.1

* just added files











lib/src/spinner_item_picker_widget.dart
// ignore_for_file: no_leading_underscores_for_local_identifiers, prefer_const_constructors_in_immutables

import 'package:flutter/material.dart';

class SpinnerItemPicker<T> extends StatefulWidget {
  // Initialize parameters for the item picker
  final int initSelectedIndex; // Initial selected item index
  final List<T> items; // List of items should be shown
  final Widget Function(dynamic item) selectedItemToWidget; // selected item to Widget function
  final Widget Function(dynamic item) nonSelectedItemToWidget; // non-selected item to Widget function
  final double height; // Height of the widget
  final double width; // Width of the widget
  final double itemHeight; // Height of individual items
  final Color spinnerBgColor; // Background color of the widget
  final void Function(dynamic item) onSelectedItemChanged; // Callback for value selection

  SpinnerItemPicker({
    Key? key,
    this.initSelectedIndex = 0,
    required this.items,
    required this.selectedItemToWidget,
    required this.nonSelectedItemToWidget,
    required this.height,
    required this.width,
    required this.itemHeight,
    required this.onSelectedItemChanged,
    required this.spinnerBgColor,
  })  : assert(items.isNotEmpty, "[items] couldn't be an empty list!"),
        assert(initSelectedIndex < items.length, "provided items not included initSelectedIndex"),
        super(key: key);

  @override
  State<SpinnerItemPicker> createState() => _SpinnerItemPickerState();
}

// Define the state for the ItemElementPicker widget
class _SpinnerItemPickerState<T> extends State<SpinnerItemPicker<T>> {
  late FixedExtentScrollController scrollController;
  late int _selectedIndex;

  @override
  void initState() {
    // Initialize state variables and scroll controller
    setState(() {
      _selectedIndex = widget.initSelectedIndex;
      scrollController = FixedExtentScrollController(initialItem: _selectedIndex);
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // Build the item picker layout
    return Container(
      height: widget.height,
      width: widget.width,
      decoration: BoxDecoration(
        color: widget.spinnerBgColor,
        borderRadius: BorderRadius.circular(10),
      ),
      child: ListWheelScrollView.useDelegate(
        controller: scrollController,
        itemExtent: widget.itemHeight, // Height of each item in the picker
        physics: const FixedExtentScrollPhysics(),
        childDelegate: ListWheelChildBuilderDelegate(
          builder: (context, index) {
            final wrappedIndex = index % widget.items.length; // Wrap around the values
            final wrappedItem = widget.items[wrappedIndex];

            final isSelectedThis = wrappedIndex == _selectedIndex;
            final wrappedItemWidget =
                isSelectedThis ? widget.selectedItemToWidget(wrappedItem) : widget.nonSelectedItemToWidget(wrappedItem);

            return Center(child: wrappedItemWidget);
          },
        ),
        onSelectedItemChanged: (index) {
          final wrappedIndex = index % widget.items.length; // Wrap around the values

          setState(() {
            _selectedIndex = wrappedIndex;
          });
          widget.onSelectedItemChanged(widget.items[_selectedIndex]); // Notify the parent about the value change
        },
      ),
    );
  }
}


lib/src/spinner_time_picker_widget.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'package:flutter/material.dart';

import 'spinner_item_picker_widget.dart';

// Define a StatefulWidget for a custom time picker widget
class SpinnerTimePicker extends StatefulWidget {
  // Initialize parameters for the time picker
  final TimeOfDay initTime; // Initial time value
  final bool is24HourFormat; // Indicates if the time format is 24-hour
  final double spinnerHeight; // Height of the widget
  final double spinnerWidth; // Width of the widget
  final double elementsSpace; // Space between hour and minute pickers
  final double digitHeight; // Height of individual time elements
  final Color spinnerBgColor; // Background color of the widget
  final TextStyle selectedTextStyle; // Text style for selected time elements
  final TextStyle nonSelectedTextStyle; // Text style for non-selected time elements
  final void Function(TimeOfDay selected) onChangedSelectedTime; // Callback for time selection

  const SpinnerTimePicker({
    Key? key,
    required this.initTime,
    required this.is24HourFormat,
    required this.spinnerHeight,
    required this.spinnerWidth,
    required this.elementsSpace,
    required this.digitHeight,
    required this.spinnerBgColor,
    required this.selectedTextStyle,
    required this.nonSelectedTextStyle,
    required this.onChangedSelectedTime,
  }) : super(key: key);

  @override
  State<SpinnerTimePicker> createState() => _SpinnerTimePickerState();
}

// Define the state for the SpinnerTimePicker widget
class _SpinnerTimePickerState extends State<SpinnerTimePicker> {
  DayPeriod selectedDayPeriod = DayPeriod.am; // Selected AM/PM period
  int selectedHour = 0; // Selected hour value
  int selectedMinute = 0; // Selected minute value

  // Options for AM and PM periods
  final _dayPeriodOptions = const [DayPeriod.am, DayPeriod.pm];

  @override
  void initState() {
    // Initialize state variables based on the initial time
    final selectedTime = widget.initTime;
    setState(() {
      selectedDayPeriod = selectedTime.period;
      selectedHour = !widget.is24HourFormat && selectedDayPeriod == DayPeriod.pm && selectedTime.hour != 12
          ? selectedTime.hour - 12
          : selectedTime.hour;
      selectedMinute = selectedTime.minute;
    });
    super.initState();
  }

  // Get a list indicating which day period is selected
  List<bool> get _isSelectedDayPeriod {
    return switch (selectedDayPeriod) {
      DayPeriod.am => [true, false], // AM is selected
      DayPeriod.pm => [false, true], // PM is selected
    };
  }

  @override
  Widget build(BuildContext context) {
    // Build the time picker layout
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      textDirection: TextDirection.ltr,
      children: [
        _hourPicker(),
        _timeSeparator(context),
        _minutePicker(),
        if (!widget.is24HourFormat) SizedBox(width: 0.7 * widget.elementsSpace),
        if (!widget.is24HourFormat) _dayPeriodSelector(),
      ],
    );
  }

  // Build the day period selector toggle buttons
  ToggleButtons _dayPeriodSelector() {
    return ToggleButtons(
      isSelected: _isSelectedDayPeriod,
      direction: Axis.vertical,
      onPressed: (index) {
        setState(() {
          selectedDayPeriod = _dayPeriodOptions[index];
        });
        setSelectedTime();
      },
      children: _dayPeriodOptions.map((option) => Text(option.name.toUpperCase())).toList(),
    );
  }

  // Build the minute picker
  SpinnerItemPicker _minutePicker() {
    final items = List.generate(60, (index) => index);
    return SpinnerItemPicker(
      items: items,
      initSelectedIndex: items.indexOf(selectedMinute),
      selectedItemToWidget: (item) {
        return Text(
          item.toString().padLeft(2, "0"), // Display with leading zero
          style: widget.selectedTextStyle,
        );
      },
      nonSelectedItemToWidget: (item) {
        return Text(
          item.toString().padLeft(2, "0"), // Display with leading zero
          style: widget.nonSelectedTextStyle,
        );
      },
      height: widget.spinnerHeight,
      width: widget.spinnerWidth,
      itemHeight: widget.digitHeight,
      spinnerBgColor: widget.spinnerBgColor,
      onSelectedItemChanged: (value) {
        setState(() {
          selectedMinute = value;
        });
        setSelectedTime();
      },
    );
  }

  // Build the time separator between hour and minute pickers
  SizedBox _timeSeparator(BuildContext context) {
    return SizedBox(
      width: widget.elementsSpace,
      child: Center(
        child: Text(
          ':',
          style: TextStyle(fontSize: 23, color: Theme.of(context).colorScheme.primary),
        ),
      ),
    );
  }

  // Build the hour picker
  SpinnerItemPicker _hourPicker() {
    final items = List.generate(widget.is24HourFormat ? 24 : 12, (index) => (widget.is24HourFormat ? index : index + 1));

    return SpinnerItemPicker(
      items: items,
      initSelectedIndex: items.indexOf(selectedHour),
      selectedItemToWidget: (item) {
        return Text(
          item.toString().padLeft(2, "0"), // Display with leading zero
          style: widget.selectedTextStyle,
        );
      },
      nonSelectedItemToWidget: (item) {
        return Text(
          item.toString().padLeft(2, "0"), // Display with leading zero
          style: widget.nonSelectedTextStyle,
        );
      },
      height: widget.spinnerHeight,
      width: widget.spinnerWidth,
      itemHeight: widget.digitHeight,
      spinnerBgColor: widget.spinnerBgColor,
      onSelectedItemChanged: (value) async {
        setState(() {
          selectedHour = value;
        });
        setSelectedTime();
      },
    );
  }

  // Update the selected time based on user choices
  void setSelectedTime() {
    final isPM = selectedDayPeriod == DayPeriod.pm;
    final is12HrFormat = !widget.is24HourFormat;
    final isNot12Hr = selectedHour != 12;
    int offset = is12HrFormat && isPM && isNot12Hr ? 12 : 0;

    widget.onChangedSelectedTime(TimeOfDay(hour: selectedHour + offset, minute: selectedMinute));
  }
}


lib/src/spinner_number_picker_widget.dart
import 'package:flutter/material.dart';

// Define a StatefulWidget for a time element picker widget
class SpinnerNumericPicker extends StatefulWidget {
  // Initialize parameters for the time element picker
  final int? initValue; // Initial value of the picker
  final int maxValue; // Maximum value of the picker
  final double height; // Height of the widget
  final double width; // Width of the widget
  final double digitHeight; // Height of individual time elements
  final TextStyle selectedTextStyle; // Text style for selected time elements
  final TextStyle nonSelectedTextStyle; // Text style for non-selected time elements
  final Color spinnerBgColor; // Background color of the widget
  final void Function(int value) onSelectedItemChanged; // Callback for value selection

  const SpinnerNumericPicker({
    Key? key,
    this.initValue,
    required this.maxValue,
    required this.height,
    required this.width,
    required this.digitHeight,
    required this.selectedTextStyle,
    required this.nonSelectedTextStyle,
    required this.onSelectedItemChanged,
    required this.spinnerBgColor,
  }) : super(key: key);

  @override
  State<SpinnerNumericPicker> createState() => _SpinnerNumericPickerState();
}

// Define the state for the TimeElementPicker widget
class _SpinnerNumericPickerState extends State<SpinnerNumericPicker> {
  late FixedExtentScrollController scrollController;
  late int _selectedValue;

  @override
  void initState() {
    // Initialize state variables and scroll controller
    setState(() {
      _selectedValue = widget.initValue ?? 0; // Set the initial value or default to 0
      scrollController = FixedExtentScrollController(initialItem: _selectedValue);
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    // Build the time element picker layout
    return Container(
      height: widget.height,
      width: widget.width,
      decoration: BoxDecoration(
        color: widget.spinnerBgColor,
        borderRadius: BorderRadius.circular(10),
      ),
      child: ListWheelScrollView.useDelegate(
        controller: scrollController,
        itemExtent: widget.digitHeight, // Height of each item in the picker
        physics: const FixedExtentScrollPhysics(),
        childDelegate: ListWheelChildBuilderDelegate(
          builder: (context, index) {
            final wrappedIndex = index % widget.maxValue; // Wrap around the values
            return Center(
              child: Text(
                wrappedIndex.toString().padLeft(2, '0'), // Display with leading zero
                style: wrappedIndex == _selectedValue ? widget.selectedTextStyle : widget.nonSelectedTextStyle,
              ),
            );
          },
        ),
        onSelectedItemChanged: (index) {
          setState(() {
            _selectedValue = index % widget.maxValue;
          });
          widget.onSelectedItemChanged(_selectedValue); // Notify the parent about the value change
        },
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_spinner_time_picker/flutter_spinner_time_picker.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Time Picker Example',
      darkTheme: ThemeData(
        primarySwatch: Colors.deepPurple,
        useMaterial3: true,
        brightness: Brightness.dark,
      ),
      theme: ThemeData(
        primarySwatch: Colors.deepPurple,
        useMaterial3: true,
        brightness: Brightness.dark,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TimeOfDay selectedTime = TimeOfDay.now();

  void _showTimePicker() async {
    final pickedTime = await showSpinnerTimePicker(
      context,
      initTime: selectedTime,
      is24HourFormat: false,
    );

    if (pickedTime != null) {
      setState(() {
        selectedTime = pickedTime;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Time Picker Example'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'Selected Time:',
              style: TextStyle(fontSize: 18),
            ),
            const SizedBox(height: 10),
            Text(
              '${selectedTime.hour}:${selectedTime.minute.toString().padLeft(2, '0')}',
              style: const TextStyle(fontSize: 30, fontWeight: FontWeight.bold),
            ),
            const SizedBox(height: 20),
            FilledButton(
              onPressed: _showTimePicker,
              child: const Text('Pick a Time'),
            ),
          ],
        ),
      ),
    );
  }
}
